controllers/replica_calculator.go

```diff
@@ -212,62 +215,117 @@ func (c *ReplicaCalculator) GetResourceReplicas(logger logr.Logger, target *auto
+
+	setCondition(wpa, v1alpha1.WatermarkPodAutoscalerStatusAboveHighWatermark, corev1.ConditionFalse, aboveHighWatermarkReason, aboveHighWatermarkAllowedMessage)
+	setCondition(wpa, v1alpha1.WatermarkPodAutoscalerStatusBelowLowWatermark, corev1.ConditionFalse, belowLowWatermarkReason, belowLowWatermarkAllowedMessage)
+	setCondition(wpa, v1alpha1.WatermarkPodAutoscalerStatusConvergeToHighWatermark, corev1.ConditionFalse, convergeToHighWatermarkAllowedMessage, convergeToHighWatermarkAllowedReason)
 
-		setCondition(wpa, v1alpha1.WatermarkPodAutoscalerStatusAboveHighWatermark, corev1.ConditionFalse, aboveHighWatermarkReason, aboveHighWatermarkAllowedMessage)
-		setCondition(wpa, v1alpha1.WatermarkPodAutoscalerStatusBelowLowWatermark, corev1.ConditionFalse, belowLowWatermarkReason, belowLowWatermarkAllowedMessage)
+	// returning the currentReplicas instead of the count of healthy ones to be consistent with the upstream behavior.
+	return currentReplicas
+}
+
+// tryToConvergeToHw will try to make the replicaCount slowly converge to high watermark but make sure that it does
+// not trigger a downscale that would put usage above HW
+func tryToConvergeToHw(logger logr.Logger, currentReplicas, currentReadyReplicas int32, wpa *v1alpha1.WatermarkPodAutoscaler, adjustedUsage float64, lowMark, highMark *resource.Quantity) (replicaCount int32) {
+
+	downScaleBy := *wpa.Spec.ReplicaScalingAbsoluteModulo
+	currentReplicasAfterDownscale := math32.Max(currentReplicas-downScaleBy, 1)
+	currentReadyReplicasAfterDownscale := math32.Max(currentReadyReplicas-downScaleBy, 0)
 
-		// returning the currentReplicas instead of the count of healthy ones to be consistent with the upstream behavior.
-		return currentReplicas, utilizationQuantity.MilliValue()
+	adjustedUsageAfterDownscale := math.Ceil(adjustedUsage * float64(currentReadyReplicas) / float64(currentReadyReplicasAfterDownscale))
+
+	if int64(adjustedUsageAfterDownscale) > highMark.MilliValue() {
+		// This would result in a downscale, give up
+		logger.Info("Scaling down would likely make usage go above HW", "replicaCount", replicaCount, "currentReadyReplicas", currentReadyReplicas, "currentReadyReplicasAfterDownscale", currentReadyReplicasAfterDownscale, "adjustedUsageAfterDownscale", adjustedUsageAfterDownscale, "highMark", highMark.MilliValue())
+		return getReplicaCountWithinBounds(currentReplicas, wpa)
 	}
 
-	restrictedScaling.With(labelsWithReason).Set(0)
+	setCondition(wpa, v1alpha1.WatermarkPodAutoscalerStatusAboveHighWatermark, corev1.ConditionFalse, aboveHighWatermarkReason, aboveHighWatermarkAllowedMessage)
+	setCondition(wpa, v1alpha1.WatermarkPodAutoscalerStatusBelowLowWatermark, corev1.ConditionFalse, belowLowWatermarkReason, belowLowWatermarkAllowedMessage)
+	setCondition(wpa, v1alpha1.WatermarkPodAutoscalerStatusConvergeToHighWatermark, corev1.ConditionTrue, convergeToHighWatermarkAllowedMessage, convergeToHighWatermarkAllowedReason)
```

Comment:
I think if we are in stable regime (i.e. between bounds) and we optimize to converge towards a watermark we should not affect the scaling windows, because we are not scaling as a result of the external signal but from our internal logic. 

I think we should improve the logic in the [`shouldScale`](https://github.com/DataDog/watermarkpodautoscaler/blob/main/controllers/watermarkpodautoscaler_controller.go#L356-L414) method to take this into account both for:
- Stable regime: Circumvent (?) `shouldScale` to have the adjustedUsage converge towards the HW with no restrictions.
- Unstable regime: A previous optimization should not prevent us from scaling how we would have if we had not used this feature. Thus, we do not affect the specific parts of the status used for the window calculation.